### PR TITLE
feat: resources with `handle` method can substitute for function literals

### LINKED ISSUE #1803: `api.get` accepts any inflight closures
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

```
bring cloud;

let api = new cloud.Api();
api.get("/hello", inflight (x: num): bool => {
  print("hi");
});
```

### I expected this:

An error saying that the expected type is `inflight (cloud.ApiRequest): cloud.ApiResponse` and not `inflight (num): bool`.

### Instead, this happened:

No compiler error.

### Component:

Compiler, SDK

### Wing Version:

0.5.126

### Node.js Version:

18.12.1

### Platform(s):

_No response_

### Anything else?

This issue applies to all preflight APIs that expect an inflight, like `api.get`, `queue.on_message`, `new cloud.Function(..)` and so on.

This happens right now because to make the compiler work before we had interfaces, we made a simplifying assumption that any preflight method/constructor that requires an inflight will have the parameter typed as "inflight (any): any" instead of a more specific inflight type.

The solution is probably to use interfaces:

https://github.com/winglang/wing/blob/9d9e15f45969f203d0801d2d2dcd0fd123596854/libs/wingsdk/src/target-tf-aws/queue.ts#L43-L46

### LINKED ISSUE #1816: Class instance cannot be assigned to interface
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

```
bring cloud;

resource Foo impl cloud.IFunctionHandler {
  init() {}
  inflight handle(message: str): str {
    return "hello world!";
  }
}

let x: cloud.IFunctionHandler = new Foo();
```

### I expected this:

It compiles

### Instead, this happened:

```
Expected type to be "IFunctionHandler", but got "Foo" instead
```

### Component:

Compiler

### Wing Version:

0.5.126

### Node.js Version:

18.12.1

### Platform(s):

MacOS

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Chriscbr**
With this PR, it's now possible to use resources in place of inflight literals as long as they implement an interface with an inflight method named "handle" with a matching type. For example:

```js
bring cloud;

resource Foo impl cloud.IFunctionHandler {
  init() {}
  inflight handle(message: str): str {
    return "hello world!";
  }
}

let fn = new cloud.Function(new Foo());

new cloud.Function(inflight () => {
  assert(fn.invoke("test") == "hello world!");
}) as "test";
```

This PR doesn't implement any changes to how code generation is performed, but this change makes progress towards removing instances of `core.Inflight` from the SDK (prereq for #1448).

Misc:
- Unify the naming around "phase" in the compiler

Fixes #1816
Fixes #1803

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -354,14 +365,11 @@ impl Subtype for Type {
 
 				// In this section, we check if the parameter types are not subtypes of each other, then this is not a subtype.
 
-				// Check that this function has at least as many required parameters as the other function requires
-				if l0.min_parameters() < r0.min_parameters() {
-					return false;
-				}
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:360]: I removed this for now since it was causing issues with type checking against interfaces... for now I've left type checking for matching # of params as a future TODO. Related: #1678

File: libs/wingc/src/type_check.rs [hunk 2]
```diff
@@ -384,7 +392,7 @@ impl Subtype for Type {
 					// (Cat) => void is not a subtype of (Animal) => void
 					// but (Animal) => void is a subtype of (Cat) => void
 					// see https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
-					if l.is_strict_subtype_of(&r) {
+					if !r.is_subtype_of(l) {
 						return false;
 					}
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:397]: Write unit tests for this

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 3]
```diff
@@ -138,14 +138,25 @@ exports[`capture_reassignable.w > stderr 1`] = `
 `;
 
 exports[`cloud_function_expects_inflight.w > stderr 1`] = `
-"error: Expected type to be \\"inflight (any): any\\", but got \\"preflight (str): str\\" instead
+"error: Expected type to be \\"IFunctionHandler\\", but got \\"preflight (str): str\\" instead
```
Comment (Chriscbr) [inline tools/hangar/__snapshots__/invalid.ts.snap:141]: This is a regression in error messages - not sure the best way to handle this yet

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -341,6 +341,17 @@ impl Subtype for Type {
 				// TODO: Hack to make anything's compatible with all other types, specifically useful for handling core.Inflight handlers
 				true
 			}
+			(Self::Function(_), Self::Interface(r0)) => {
+				// TODO: Hack to make functions compatible with interfaces
+				// Remove this after https://github.com/winglang/wing/issues/1448
+				// Compare the function to a method on the interface named "handle" if it exists
+				if let Some(method) = r0.get_env().try_lookup("handle", None) {
+					let method = method.as_variable().unwrap();
+					self.is_subtype_of(&*method.type_)
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:350]: Check that the method is an inflight method (otherwise return false)

File: libs/wingc/src/type_check.rs [hunk 5]
```diff
@@ -2922,4 +2951,126 @@ mod tests {
...
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let str_fn = Type::Function(FunctionSignature {
+			parameters: vec![string],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:3011]: something like this might be clearer (also for all the other tests)
```suggestion
        let sig = FunctionSignature {
			parameters: vec![num],
			return_type: void,
			phase: Phase::Inflight,
			js_override: None,
		};
		let num_fn = Type::Function(sig);
		let str_fn = Type::Function(FunctionSignature {
			parameters: vec![string], ..sig.clone()
		});
```

File: libs/wingc/src/type_check.rs [hunk 6]
```diff
@@ -2922,4 +2951,126 @@ mod tests {
...
+
+		// functions of incompatible return types are not subtypes of each other
+		assert!(!returns_num.is_subtype_of(&returns_str));
+		assert!(!returns_str.is_subtype_of(&returns_num));
+
+		// functions with specific return types are subtypes of functions with void return type
+		assert!(returns_num.is_subtype_of(&returns_void));
+		assert!(returns_str.is_subtype_of(&returns_void));
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:3048]: I'm not sure I approve of this relationship needed :)  Why is this needed?

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:350]: Since this is probably a temp hack, I'd do something along the lines of this to hack the `this` first arg and not change the function comparison below.
```suggestion
					let temp_fn = Self::Function(FunctionSignature { parameters: (vec![other]).concat(l0.parameters.clone()), ..l0.clone()});
					temp_fn.is_subtype_of(&*method.type_)
```

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs:360]: lets not do this change. We'll fix this to something correct in #1678, but I don't want to hurt function comparison for this. See comment above.

File: libs/wingc/src/type_check.rs [hunk 7]
```diff
@@ -727,6 +735,7 @@ impl TypeRef {
 	pub fn is_capturable(&self) -> bool {
 		match **self {
 			Type::Resource(_) => true,
+			Type::Interface(_) => true,
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:738]: Why do we need this? It's not strictly correct.

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: Looks like yes

File: examples/tests/invalid/impl_interface.w [hunk 8]
```diff
@@ -7,8 +7,12 @@ resource A impl cloud.IQueueOnMessageHandler {
...
   init() {}
-  inflight handle() {
-    // Error: Expected type to be "inflight (str): void", but got "inflight (): void" instead
+  inflight handle(x: num) {
+    // Error: Expected type to be "inflight (str): void", but got "inflight (num): void" instead
     return;
   }
 }
+
+resource C impl cloud.Bucket {
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply examples/tests/invalid/impl_interface.w:16]: Added

File: libs/wingc/src/type_check.rs (hunk 6 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:3057]: The reason I have this in place is so that both of these kinds of existing usages in tests will work:

```js
let fn1 = new cloud.Function(inflight (m: str) {
  print("im not returning anything");
});

let fn2 = new cloud.Function(inflight (m: str): str {
  return "Hello, ${m}!";
});

let usage = inflight () => {
  fn1.invoke("");
  let result = fn2.invoke("");
};
```

We don't have generics, so we're kind of in a tough place as far as offering flexible parameter and return types, since `fn.invoke()` needs to always return the SAME type.

The alternative to this (if we don't want this special return-type subtyping rule) is that every `cloud.Function` must have the same return type, which would have to be `str` or `Json`. We could change it to `str` or `Json` but it would be a breaking change at the moment.

Let's open a separate issue about this because I don't think it's core to this PR

Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:3057]: Related: https://github.com/winglang/wing/issues/1843

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: Done

File: libs/wingc/src/type_check.rs [hunk 9]
```diff
@@ -2103,7 +2125,13 @@ impl<'a> TypeChecker<'a> {
 				for interface in implements.iter() {
 					let interface_type =
 						resolve_user_defined_type(interface, env, stmt.idx).unwrap_or_else(|e| self.type_error(e));
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: 👍

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: Ok, added it back

File: libs/wingc/src/type_check.rs (hunk 7 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:764]: This is necessary so that the resources typed as interfaces can be used inflight. Here's an example:

```js
bring cloud;

resource A {
  fn: cloud.IFunctionHandler;
  init(fn: cloud.IFunctionHandler) {
    this.fn = fn;
  }
  inflight do_stuff(message: str): str {
    this.fn.handle(message);
  }
}

resource Greeter impl cloud.IFunctionHandler {
  init() {}
  inflight handle(message: str): str {
    return "Hello ${message}";
  }
}

let fn = new A(new Greeter());
```

Perhaps we need some way to distinguish between interfaces on resources and interfaces on non-resources?

Comment (Chriscbr) [reply libs/wingc/src/type_check.rs]: OK, I've added a check. It's hacky since I couldn't see an easy way to check for type identity but it should be good enough for now

**Comment by Chriscbr** [comment]
@yoav-steinberg Ready for another pass

File: libs/wingc/src/type_check.rs [hunk 10]
```diff
@@ -341,6 +341,39 @@ impl Subtype for Type {
...
+				}
+
+				// Next, compare the function to a method on the interface named "handle" if it exists
+				if let Some(method) = r0.get_env().try_lookup("handle", None) {
+					let method = method.as_variable().unwrap();
+					if method.phase != Phase::Inflight {
+						return false;
+					}
+
+					// hallucinate a "this" parameter for the function, since the
```
Comment (eladb) [inline libs/wingc/src/type_check.rs:360]: Very zeitgeist

File: libs/wingc/src/type_check.rs [hunk 11]
```diff
@@ -354,8 +387,9 @@ impl Subtype for Type {
 
 				// In this section, we check if the parameter types are not subtypes of each other, then this is not a subtype.
 
-				// Check that this function has at least as many required parameters as the other function requires
-				if l0.min_parameters() < r0.min_parameters() {
+				// Check that this function has at most as many required parameters as the other function requires
+				// if it doesn't, we know it's not a subtype
+				if l0.min_parameters() > r0.min_parameters() {
```
Comment (eladb) [inline libs/wingc/src/type_check.rs:392]: Shouldn't this be part of how we evaluate subtypes for functions?

File: libs/wingc/src/type_check.rs [hunk 12]
```diff
@@ -738,8 +772,23 @@ impl TypeRef {
 	}
 
 	pub fn is_capturable(&self) -> bool {
-		match **self {
+		match &**self {
 			Type::Resource(_) => true,
+			Type::Interface(iface) => {
+				// TODO: This should check that the interface extends `IResource` from
+				// the SDK, not just any interface with the name `IResource`
+				if iface.name.name == "IResource"
```
Comment (eladb) [inline libs/wingc/src/type_check.rs:780]: Maybe extract to a helper like `iface.is_resource()`

File: libs/wingc/src/type_check.rs (hunk 11 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:392]: That's what this code is

Comment (eladb) [reply libs/wingc/src/type_check.rs:410]: Okay oops. Nice I see!

Awesome!

File: libs/wingc/src/type_check.rs [hunk 13]
```diff
@@ -207,6 +207,24 @@ pub struct Interface {
 	pub env: SymbolEnv,
 }
 
+impl Interface {
+	fn is_resource(&self) -> bool {
+		// TODO: This should check that the interface extends `IResource` from
+		// the SDK, not just any interface with the name `IResource`
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:213]: The way to do this it to lookup the `cloud.IResource` name in the std lib namespace. And then compare the pointers (since there should be only one such interface imported). But I'm fine with making this a future task.

File: libs/wingc/src/type_check.rs [hunk 14]
```diff
@@ -2988,6 +3030,15 @@ mod tests {
 		assert!(!Phase::Inflight.is_subtype_of(&Phase::Preflight));
 	}
 
+	fn make_function(params: Vec<TypeRef>, ret: TypeRef, phase: Phase) -> Type {
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:3033]: probably need this to avoid a dead code warning
```suggestion
	#[cfg(test)]
	fn make_function(params: Vec<TypeRef>, ret: TypeRef, phase: Phase) -> Type {
```

**Review by yoav-steinberg — APPROVED** [review]
really nice work

File: libs/wingc/src/type_check.rs (hunk 14 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:3033]: `#[cfg(test)]` is already applied to the module it's in (`mod tests`) so I haven't gotten any dead code warnings

File: libs/wingc/src/type_check.rs (hunk 13 above)
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:213]: Yeah, I wasn't sure how to do a lookup from here. One option might be to move this method so it exists on the `Types` struct instead, but that feels a bit weird, so I was also thinking of simply adding `is_resource` as a `bool` field on `Interface`

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 3 above)
Comment (Chriscbr) [reply tools/hangar/__snapshots__/invalid.ts.snap]: OK, figured out a way to get a reasonable-ish experience -- if there is an inflight method named "handle" on the interface, it will be displayed in parentheses
